packages/govuk-frontend/src/govuk/components/header/header.mjs

```diff
@@ -52,22 +52,30 @@ export class Header extends GOVUKFrontendComponent {
     }
 
     this.$module = $module
-    this.$menuButton = $module.querySelector('.govuk-js-header-toggle')
-    this.$menu =
-      this.$menuButton &&
-      $module.querySelector(
-        `#${this.$menuButton.getAttribute('aria-controls')}`
-      )
-
-    if (
-      !(
-        this.$menuButton instanceof HTMLElement ||
-        this.$menu instanceof HTMLElement
-      )
-    ) {
+    const $menuButton = $module.querySelector('.govuk-js-header-toggle')
+
+    // Headers don't necessarily have a navigation.
+    // When they don't, the menu toggle won't be rendered
+    // by our macro (or may be omitted when writing plain HTML)
+    if (!($menuButton instanceof HTMLElement)) {
       return this
     }
```

Comment:
Hmm, the plain HTML comment has caught my eye

Due to progressive enhancement we fall back to the menu being open

But remove the button and invalid navigation menus get a `return this` free pass

(We should still check them)

Reply:
Do you mean a Header having a menu but not a toggle, which would fall into this `return this`?

Reply:
Looked a bit more into it: I think with a missing button, we would not have been checking the type of the `$menu` anyway due to it begin `false` because `this.$menuButton` being falsey anyway (first line of the table).

Slightly different matter if `$menuButton` was just not an instance of `HTMLElement`. If `$menu` was an instance of `HTMLElement` we wouldn't be return early while we should have (1).

Digging a bit further, I think there was a few instance where the type of `$menu` wasn't checked where it should have been because of the `||` not running the second clause if the first is truthy (2).

$menuButton | $menu | $menuButton instanceof HTMLElement | $menu instanceof HTMLElement | returning early
-- | -- | -- | -- | --
`null` from not being on the page | Can only be `false` from `this.$menuButton` being falsey | FALSE | FALSE | Yes
an `HTMLElement` | an `HTMLElement` | TRUE | Not run | No (2)
not an `HTMLElement` | an `HTMLElement` | FALSE | TRUE | No (1)
an `HTMLElement` | not an `HTMLElement` | TRUE | Not run | No (2)
not an `HTMLElement` | not an `HTMLElement` | FALSE | FALSE | Yes

I _think_ the following implementation limits the free pass for `$menu` to when `$menuButton` is `null` (like in `v4`) while making sure its type gets checked when `$menuButton` is an `HTMLElement` (2). No checks when `$menuButton` is not an HTMLElement, but the component would not carry further as it mistakenly does at the moment.

```
    // Headers don't necessarily have a navigation.
    // When they don't, the menu toggle won't be rendered
    // by our macro (or may be omitted when writing plain HTML)
    if (!$menuButton) {
      return this
    }

    const $menu = $module.querySelector(
      `#${$menuButton.getAttribute('aria-controls')}`
    )

    if (!($menuButton instanceof HTMLElement)) {
      throw new ElementError($menuButton, {
        componentName: 'Header',
        identifier: 'Menu toggle'
      })
    }

    if (!($menu instanceof HTMLElement)) {
      throw new ElementError($menu, {
        componentName: 'Header',
        identifier: 'Menu'
      })
    }

```

Reply:
Thanks for the write up 😎

I'm afraid the free pass is still there, but as:

```mjs
if (!$menuButton) {
  return this
}
```

i.e. By removing the `$menuButton` I can fall back to an always-open `<svg>` menu without errors

Can we stick with the usual approach of:

1. Run a selector
2. Check if missing or wrong type
3. Add to class field after checks pass

```mjs
// 1. Run selector
const $menu = $module.querySelector('#navigation')
    
// 2. Check if missing or wrong type
if (!($menu instanceof HTMLElement)) {
  throw new ElementError($menu, {
    componentName: 'Header',
    identifier: 'Menu'
  })
}
```

Another error would be `$menuButton.getAttribute('aria-controls')` not matching `navigation`

Reply:
Yeah, that would work if we add a check for the link of the `$menu.id` and `$menuButton.getAttribute('aria-controls')`.

Reply:
@romaricpascal Brill, glad that did the trick

Yeah please, it's commented out in https://github.com/alphagov/govuk-frontend/pull/4206/commits/7660f032b83e54d741ab4cb9e707c65648c456ba (commit after) but would be crucial. Hopefully easier to string a nice error message together since we merged https://github.com/alphagov/govuk-frontend/pull/4219/commits/8965564d8c087ada039adf008276da99668bd8c0 too

```mjs
if (!$menu.id || $menu.id !== $menuButton.getAttribute('aria-controls')) {
  throw new ElementError($menuButton, {
    componentName: 'Header',
    identifier: '$menuButton["aria-controls"]',
    expectedType: '???'
  })
}
```